Peers module unit tests are failing randomly

Branch: 0.9.11

Reference: https://jenkins.lisk.io/job/lisk-core/job/1024_refactor-transaction-pool/3/execution/node/227/log/

```
    ping
      ✓ should accept peer with public ip
    onBlockchainReady
[...]
>> AssertionError: expected false to be truthy
```

--- a/test/common/helpers.js
+++ b/test/common/helpers.js
@@ -20,6 +20,13 @@ var stripTransactionsResults = function (results) {
 	};
 };
 
+var randomInt = function (min, max) {
+	min = Math.ceil(min);
+	max = Math.floor(max);
+	return Math.floor(Math.random() * (max - min + 1)) + min;
+};
+
 module.exports = {
-	stripTransactionsResults: stripTransactionsResults
+	stripTransactionsResults: stripTransactionsResults,
+	randomInt: randomInt
 };

--- a/test/common/initModule.js
+++ b/test/common/initModule.js
@@ -124,7 +124,7 @@ var modulesLoader = new function () {
 				this.getDbConnection(waterCb);
 			}.bind(this),
 			function (db, waterCb) {
-				scope = _.merge(this.scope, {db: db}, scope);
+				scope = _.assign({}, this.scope, {db: db}, scope);
 				async.reduce(logic, {}, function (memo, logicObj, mapCb) {
 					var name = _.keys(logicObj)[0];
 					return this.initLogic(logicObj[name], scope, function (err, initializedLogic) {
@@ -134,7 +134,7 @@ var modulesLoader = new function () {
 				}.bind(this), waterCb);
 			}.bind(this),
 			function (logic, waterCb) {
-				scope = _.merge(this.scope, {logic: logic}, scope);
+				scope.logic = _.merge(logic, scope.logic);
 				async.reduce(modules, {}, function (memo, moduleObj, mapCb) {
 					var name = _.keys(moduleObj)[0];
 					return this.initModule(moduleObj[name], scope, function (err, module) {

--- a/test/common/objectStubs.js
+++ b/test/common/objectStubs.js
@@ -1,6 +1,6 @@
 'use strict';
 
-var sinon = require('sinon');
+var randomstring = require('randomstring');
 
 var randomPeer = {
 	'broadhash': '198f2b61a8eb95fbeed58b8216780b68f697f26b849acf00c8c93bb9b24f783d',
@@ -13,6 +13,24 @@ var randomPeer = {
 	'version': '0.0.0'
 };
 
+function generateRandomActivePeer () {
+	var randomDigits = function (length) {
+		return randomstring.generate({charset: 'numeric', length: length});
+	};
+	return {
+		'broadhash': randomstring.generate(64),
+		'dappid': null,
+		'height': randomDigits(4),
+		'ip': randomDigits(3) + '.' + randomDigits(3) + '.' + randomDigits(3) + '.' + randomDigits(3),
+		'os': randomstring.generate(10),
+		'port': randomstring.generate({charset: 'numeric', length: 4}),
+		'state': 2,
+		'nonce': randomstring.generate(16),
+		'version': randomDigits(1) + '.' + randomDigits(1) + '.' + randomDigits(1)
+	};
+}
+
 module.exports = {
-	randomPeer: randomPeer
+	randomPeer: randomPeer,
+	generateRandomActivePeer: generateRandomActivePeer
 };

--- a/test/unit/modules/peers.js
+++ b/test/unit/modules/peers.js
@@ -83,7 +83,7 @@ describe('peers', function () {
 		describe('when logic.peers.list returns no records', function () {
 
 			before(function () {
-				peersLogicMock.list.resetHistory();
+				peersLogicMock.list.reset();
 				peersLogicMock.list.returns([]);
 			});
 
@@ -317,7 +317,7 @@ describe('peers', function () {
 		});
 
 		beforeEach(function () {
-			peersLogicMock.upsert.resetHistory();
+			peersLogicMock.upsert.reset();
 			peersLogicMock.upsert.returns(validUpsertResult);
 			updateResult = peers.update(validPeer);
 		});
@@ -363,7 +363,7 @@ describe('peers', function () {
 		});
 
 		beforeEach(function () {
-			peersLogicMock.remove.resetHistory();
+			peersLogicMock.remove.reset();
 			peersLogicMock.remove.returns(validLogicRemoveResult);
 			removeResult = peers.remove(validIp, validPort);
 		});

--- a/test/unit/modules/peers.js
+++ b/test/unit/modules/peers.js
@@ -1,52 +1,50 @@
 'use strict';
 
-var chai = require('chai');
 var expect = require('chai').expect;
-var express = require('express');
 var sinon = require('sinon');
-var randomString = require('randomstring');
-var _  = require('lodash');
+var randomstring = require('randomstring');
+var _ = require('lodash');
 
-var config = require('../../config.json');
 var randomPeer = require('../../common/objectStubs').randomPeer;
+var randomInt = require('../../common/helpers').randomInt;
+var generateRandomActivePeer = require('../../common/objectStubs').generateRandomActivePeer;
 var modulesLoader = require('../../common/initModule').modulesLoader;
-
-var currentPeers = [];
+var constants = require('../../../helpers/constants');
 
 describe('peers', function () {
 
-	var peers, modules;
+	var peers;
+	var peersLogicMock;
+	var modules;
 
-	var NONCE = randomString.generate(16);
+	var NONCE = randomstring.generate(16);
 
-	function getPeers (cb) {
-		peers.list({broadhash: config.nethash}, function (err, __peers) {
-			expect(err).to.not.exist;
-			expect(__peers).to.be.an('array');
-			return cb(err, __peers);
-		});
-	}
+	before(function () {
+		peersLogicMock = {
+			create: sinon.spy(),
+			exists: sinon.stub(),
+			get: sinon.stub(),
+			list: sinon.stub(),
+			upsert: sinon.stub(),
+			remove: sinon.stub()
+		};
+	});
 
 	before(function (done) {
 		modulesLoader.initAllModules(function (err, __modules) {
-			if (err) {
-				return done(err);
-			}
 			peers = __modules.peers;
 			modules = __modules;
 			peers.onBind(__modules);
-			done();
-		}, {nonce: NONCE});
-	});
-
-	beforeEach(function (done) {
-		getPeers(function (err, __peers) {
-			currentPeers = __peers;
-			done();
+			done(err);
+		}, {
+			nonce: NONCE,
+			logic: {
+				peers: peersLogicMock
+			}
 		});
 	});
 
-	describe('sandboxApi', function (done) {
+	describe('sandboxApi', function () {
 
 		it('should pass the call', function () {
 			var sandboxHelper = require('../../../helpers/sandbox.js');
@@ -57,141 +55,367 @@ describe('peers', function () {
 		});
 	});
 
+	describe('list', function () {
+
+		var listError;
+		var listResult;
+		var validOptions;
+		var randomPeers;
+
+		before(function () {
+			validOptions = {};
+			// Set TEST variable in case of public ip address gets generated
+			process.env['NODE_ENV'] = 'TEST';
+		});
+
+		after(function () {
+			process.env['NODE_ENV'] = '';
+		});
+
+		beforeEach(function (done) {
+			peers.list(validOptions, function (err, peersResult) {
+				listError = err;
+				listResult = peersResult;
+				done();
+			});
+		});
+
+		describe('when logic.peers.list returns no records', function () {
+
+			before(function () {
+				peersLogicMock.list.resetHistory();
+				peersLogicMock.list.returns([]);
+			});
+
+			it('should return an empty array', function () {
+				expect(listResult).to.be.an('array').and.to.be.empty;
+			});
+		});
+
+		describe('when logic.peers.list returns 1000 random connected peers', function () {
+
+			before(function () {
+				randomPeers = _.range(1000).map(function () {
+					return generateRandomActivePeer();
+				});
+				peersLogicMock.list.returns(randomPeers);
+			});
+
+			it('should return all 1000 peers', function () {
+				expect(listResult).be.an('array').and.have.lengthOf(100);
+			});
+
+			describe('options.limit', function () {
+
+				describe('when options.limit < 1000', function () {
+
+					var validLimit;
+
+					before(function () {
+						validLimit = randomInt(1, (1000 - 1));
+						validOptions.limit = validLimit;
+					});
+
+					afterEach(function () {
+						// List arguments are mutated - needs to be overwritten after every test
+						validOptions.limit = validLimit;
+					});
+
+					after(function () {
+						delete validOptions.limit;
+					});
+
+					it('should return up to [options.limit] results', function () {
+						expect(listResult).be.an('array').and.have.lengthOf(validLimit);
+					});
+				});
+
+				describe('when no options.limit passed', function () {
+
+					it('should return [constants.maxPeers] results', function () {
+						expect(listResult).be.an('array').and.have.lengthOf(constants.maxPeers);
+					});
+				});
+			});
+
+			describe('options.broadhash', function () {
+
+				describe('when 250 peers matching and 750 not matching broadhash', function () {
+
+					var validBroadhash;
+					var validLimit;
+
+					before(function () {
+						// Ensure that different than checking broadhashes will be generated
+						var characterNotPresentInValidBroadhash = '@';
+						validBroadhash = randomstring.generate({
+							length: 64,
+							custom: 'abcdefghijklmnopqrstuvwxyz0123456789!$&_.'
+						});
+						validOptions.broadhash = validBroadhash;
+						// 250 peers matching broadhash, next 750 with different one
+						_.range(1000).forEach(function (i) {
+							randomPeers[i].broadhash = i < 250 ? validBroadhash :
+								randomstring.generate({
+									length: 63,
+									custom: 'abcdefghijklmnopqrstuvwxyz0123456789!$&_.'
+								}) + characterNotPresentInValidBroadhash;
+						});
+					});
+
+					after(function () {
+						delete validOptions.broadhash;
+						delete validOptions.limit;
+					});
+
+					describe('when options.limit = 100', function () {
+
+						before(function () {
+							validLimit = 100;
+							validOptions.limit = validLimit;
+						});
+
+						it('should return 100 results', function () {
+							expect(listResult).be.an('array').and.have.lengthOf(100);
+						});
+
+						it('should return 100 results with the same broadhash', function () {
+							expect(listResult.filter(function (peer) {
+								return peer.broadhash === validBroadhash;
+							})).be.an('array').and.have.lengthOf(100);
+						});
+					});
+
+					describe('when options.limit = 500', function () {
+
+						before(function () {
+							validLimit = 500;
+							validOptions.limit = validLimit;
+						});
+
+						afterEach(function () {
+							// List arguments are mutated - needs to be overwritten after every test
+							validOptions.limit = validLimit;
+						});
+
+						it('should return 500 results', function () {
+							expect(listResult).be.an('array').and.have.lengthOf(500);
+						});
+
+						it('should return 250 results with the same broadhash', function () {
+							expect(listResult.filter(function (peer) {
+								return peer.broadhash === validBroadhash;
+							})).be.an('array').and.have.lengthOf(250);
+						});
+
+						it('should return 250 results with different broadhash', function () {
+							expect(listResult.filter(function (peer) {
+								return peer.broadhash !== validBroadhash;
+							})).be.an('array').and.have.lengthOf(250);
+						});
+					});
+				});
+
+				describe('when no options.limit passed', function () {
+
+					it('should return [constants.maxPeers] results', function () {
+						expect(listResult).be.an('array').and.have.lengthOf(constants.maxPeers);
+					});
+				});
+			});
+		});
+
+		describe('when logic.peers.list returns 1000 random state peers and limit = 1000', function () {
+
+			describe('options.allowedStates', function () {
+
+				var CONNECTED_STATE = 2;
+				var BANNED_STATE = 1;
+				var DISCONNECTED_STATE = 0;
+
+				before(function () {
+					validOptions.limit = 1000;
+					randomPeers = _.range(1000).map(function () {
+						var peer = generateRandomActivePeer();
+						peer.state = randomInt(DISCONNECTED_STATE, CONNECTED_STATE);
+						return peer;
+					});
+					peersLogicMock.list.returns(randomPeers);
+				});
+
+				after(function () {
+					delete validOptions.limit;
+				});
+
+				it('should return only connected peers', function () {
+					expect(_.uniqBy(listResult, 'state')).be.an('array').and.have.lengthOf(1);
+					expect(listResult[0].state).equal(CONNECTED_STATE);
+				});
+
+				describe('when options.allowedStates = [1]', function () {
+
+					before(function () {
+						validOptions.allowedStates = [1];
+					});
+
+					after(function () {
+						delete validOptions.allowedStates;
+					});
+
+					it('should return only banned peers', function () {
+						expect(_.uniqBy(listResult, 'state')).be.an('array').and.have.lengthOf(1);
+						expect(listResult[0].state).equal(BANNED_STATE);
+					});
+				});
+
+				describe('when options.allowedStates = [0]', function () {
+
+					before(function () {
+						validOptions.allowedStates = [0];
+					});
+
+					after(function () {
+						delete validOptions.allowedStates;
+					});
+
+					it('should return only disconnected peers', function () {
+						expect(_.uniqBy(listResult, 'state')).be.an('array').and.have.lengthOf(1);
+						expect(listResult[0].state).equal(DISCONNECTED_STATE);
+					});
+				});
+
+				describe('when options.allowedStates = [0, 1]', function () {
+
+					before(function () {
+						validOptions.allowedStates = [0, 1];
+					});
+
+					after(function () {
+						delete validOptions.allowedStates;
+					});
+
+					it('should return disconnected and banned peers', function () {
+						expect(_.uniqBy(listResult, 'state')).be.an('array').and.have.length.at.least(1);
+						listResult.forEach(function (state) {
+							expect(state).not.to.equal(CONNECTED_STATE);
+						});
+					});
+				});
+			});
+		});
+	});
+
 	describe('update', function () {
 
-		it('should insert new peer', function (done) {
-			peers.update(randomPeer);
+		var validPeer;
+		var updateResult;
+		var validUpsertResult;
 
-			getPeers(function (err, __peers) {
-				expect(currentPeers.length + 1).that.equals(__peers.length);
-				currentPeers = __peers;
-				var inserted = __peers.find(function (p) {
-					return p.ip + ':' + p.port === randomPeer.ip + ':' + randomPeer.port;
-				});
-				expect(inserted).to.be.an('object');
-				expect(inserted).not.to.be.empty;
-				done();
-			});
+		before(function () {
+			validUpsertResult = true;
+			validPeer = generateRandomActivePeer();
 		});
 
-		it('should update existing peer', function (done) {
-			var toUpdate = _.clone(randomPeer);
-			toUpdate.height += 1;
-			peers.update(toUpdate);
-
-			getPeers(function (err, __peers) {
-				expect(currentPeers.length).that.equals(__peers.length);
-				currentPeers = __peers;
-				var updated = __peers.find(function (p) {
-					return p.ip + ':' + p.port === randomPeer.ip + ':' + randomPeer.port;
-				});
-				expect(updated).to.be.an('object');
-				expect(updated).not.to.be.empty;
-				expect(updated.ip + ':' + updated.port).that.equals(randomPeer.ip + ':' + randomPeer.port);
-				expect(updated.height).that.equals(toUpdate.height);
-				done();
-			});
+		beforeEach(function () {
+			peersLogicMock.upsert.resetHistory();
+			peersLogicMock.upsert.returns(validUpsertResult);
+			updateResult = peers.update(validPeer);
 		});
 
-		it('should insert new peer if ip or port changed', function (done) {
-			var toUpdate = _.clone(randomPeer);
-			toUpdate.port += 1;
-			peers.update(toUpdate);
-
-			getPeers(function (err, __peers) {
-				expect(currentPeers.length + 1).that.equals(__peers.length);
-				currentPeers = __peers;
-				var inserted = __peers.find(function (p) {
-					return p.ip + ':' + p.port === toUpdate.ip + ':' + toUpdate.port;
-				});
-				expect(inserted).to.be.an('object');
-				expect(inserted).not.to.be.empty;
-				expect(inserted.ip + ':' + inserted.port).that.equals(toUpdate.ip + ':' + toUpdate.port);
-
-				toUpdate.ip = '40.40.40.41';
-				peers.update(toUpdate);
-				getPeers(function (err, __peers) {
-					expect(currentPeers.length + 1).that.equals(__peers.length);
-					currentPeers = __peers;
-					var inserted = __peers.find(function (p) {
-						return p.ip + ':' + p.port === toUpdate.ip + ':' + toUpdate.port;
-					});
-					expect(inserted).to.be.an('object');
-					expect(inserted).not.to.be.empty;
-					expect(inserted.ip + ':' + inserted.port).that.equals(toUpdate.ip + ':' + toUpdate.port);
-					done();
-				});
-			});
+		it('should call logic.peers.upsert', function () {
+			expect(peersLogicMock.upsert.calledOnce).to.be.true;
 		});
 
-		var ipAndPortPeer = {
-			ip: '40.41.40.41',
-			port: 4000
-		};
-
-		it('should insert new peer with only ip and port defined', function (done) {
-			peers.update(ipAndPortPeer);
-
-			getPeers(function (err, __peers) {
-				expect(currentPeers.length + 1).that.equals(__peers.length);
-				currentPeers = __peers;
-				var inserted = __peers.find(function (p) {
-					return p.ip + ':' + p.port === ipAndPortPeer.ip + ':' + ipAndPortPeer.port;
-				});
-				expect(inserted).to.be.an('object');
-				expect(inserted).not.to.be.empty;
-				expect(inserted.ip + ':' + inserted.port).that.equals(ipAndPortPeer.ip + ':' + ipAndPortPeer.port);
-				done();
-			});
+		it('should call logic.peers.upsert with peer', function () {
+			expect(peersLogicMock.upsert.calledWith(validPeer)).to.be.true;
 		});
 
-		it('should update peer with only one property defined', function (done) {
-			peers.update(ipAndPortPeer);
+		it('should return library.logic.peers.upsert result', function () {
+			expect(updateResult).equal(validUpsertResult);
+		});
 
-			getPeers(function (err, __peers) {
-				currentPeers = __peers;
+		describe('when peer state != 2', function () {
 
-				var almostEmptyPeer = _.clone(ipAndPortPeer);
-				almostEmptyPeer.height = 1;
+			var differentThanTwoState = 1;
 
-				peers.update(almostEmptyPeer);
-				getPeers(function (err, __peers) {
-					expect(currentPeers.length).that.equals(__peers.length);
-					var inserted = __peers.find(function (p) {
-						return p.ip + ':' + p.port === ipAndPortPeer.ip + ':' + ipAndPortPeer.port;
-					});
-					expect(inserted).to.be.an('object');
-					expect(inserted).not.to.be.empty;
-					expect(inserted.ip + ':' + inserted.port).that.equals(ipAndPortPeer.ip + ':' + ipAndPortPeer.port);
-					expect(inserted.height).that.equals(almostEmptyPeer.height);
-					done();
-				});
+			before(function () {
+				validPeer.state = differentThanTwoState;
+			});
+
+			it('should call logic.peers.upsert with peer containing state = 2 anyway', function () {
+				expect(peersLogicMock.upsert.calledWith(sinon.match({state: 2}))).to.be.true;
 			});
 		});
 	});
 
 	describe('remove', function () {
 
-		before(function (done) {
-			peers.update(randomPeer);
-			done();
+		var validIp;
+		var validPort;
+		var removeResult;
+		var validLogicRemoveResult;
+
+		before(function () {
+			validLogicRemoveResult = true;
+			var validPeer = generateRandomActivePeer();
+			validIp = validPeer.ip;
+			validPort = validPeer.port;
 		});
 
-		it('should remove added peer', function (done) {
-			getPeers(function (err, __peers) {
-				currentPeers = __peers;
-				var peerToRemove = currentPeers.find(function (p) {
-					return p.ip + ':' + p.port === randomPeer.ip + ':' + randomPeer.port;
-				});
-				expect(peerToRemove).to.be.an('object').and.not.to.be.empty;
-				expect(peerToRemove.state).that.equals(2);
+		beforeEach(function () {
+			peersLogicMock.remove.resetHistory();
+			peersLogicMock.remove.returns(validLogicRemoveResult);
+			removeResult = peers.remove(validIp, validPort);
+		});
 
-				expect(peers.remove(peerToRemove.ip, peerToRemove.port)).to.be.ok;
-				getPeers(function (err, __peers) {
-					expect(currentPeers.length - 1).that.equals(__peers.length);
-					currentPeers = __peers;
-					done();
-				});
+		describe('when removable peer is frozen', function () {
+
+			var originalFrozenPeersList;
+			var loggerDebugSpy;
+
+			before(function () {
+				originalFrozenPeersList = _.assign({}, modulesLoader.scope.config.peers.list);
+				modulesLoader.scope.config.peers.list = [{
+					ip: validIp,
+					port: validPort
+				}];
+				loggerDebugSpy = sinon.spy(modulesLoader.scope.logger, 'debug');
+			});
+
+			after(function () {
+				modulesLoader.scope.config.peers.list = originalFrozenPeersList;
+				loggerDebugSpy.restore();
+			});
+
+			it('should not call logic.peers.remove', function () {
+				expect(peersLogicMock.remove.called).to.be.false;
+			});
+
+			it('should call logger.debug with message = "Cannot remove frozen peer"', function () {
+				expect(loggerDebugSpy.calledWith('Cannot remove frozen peer')).to.be.true;
+			});
+
+			it('should call logger.debug with message = [ip:port]', function () {
+				expect(loggerDebugSpy.args[0][1]).eql(validIp + ':' + validPort);
+			});
+		});
+
+		describe('when removable peer is not frozen', function () {
+
+			it('should call logic.peers.remove', function () {
+				expect(peersLogicMock.remove.calledOnce).to.be.true;
+			});
+
+			it('should call logic.peers.remove with object containing expected ip', function () {
+				expect(peersLogicMock.remove.calledWith(sinon.match({ip: validIp}))).to.be.true;
+			});
+
+			it('should call logic.peers.remove with object containing expected port', function () {
+				expect(peersLogicMock.remove.calledWith(sinon.match({port: validPort}))).to.be.true;
+			});
+
+			it('should return library.logic.peers.remove result', function () {
+				expect(removeResult).equal(validLogicRemoveResult);
 			});
 		});
 	});
@@ -263,12 +487,23 @@ describe('peers', function () {
 
 	describe('onBlockchainReady', function () {
 
+		var peersDiscoverStub;
+		var pingStub;
+
 		before(function () {
+			peersLogicMock.create = sinon.stub().returnsArg(0);
 			modules.transport.onBind(modules);
+			peersDiscoverStub = sinon.stub(peers, 'discover');
+			modulesLoader.scope.config.peers.list = [];
+			peersLogicMock.create.returnsArg(0);
+			pingStub = sinon.stub(peers, 'ping');
+		});
+
+		after(function () {
+			pingStub.restore();
 		});
 
 		it('should update peers during onBlockchainReady', function (done) {
-			sinon.stub(peers, 'discover').callsArgWith(0, null);
 			var config = require('../../config.json');
 			var initialPeers = _.clone(config.peers.list);
 			if (initialPeers.length === 0) {
@@ -277,7 +512,6 @@ describe('peers', function () {
 			peers.onBlockchainReady();
 			setTimeout(function () {
 				expect(peers.discover.calledOnce).to.be.ok;
-				peers.discover.restore();
 				done();
 			}, 100);
 		});
@@ -286,15 +520,14 @@ describe('peers', function () {
 	describe('onPeersReady', function () {
 
 		before(function () {
-			modules.transport.onBind(modules);
+			peersLogicMock.list.returns([]);
+			peers.discover = sinon.stub();
 		});
 
 		it('should update peers during onBlockchainReady', function (done) {
-			sinon.stub(peers, 'discover').callsArgWith(0, null);
 			peers.onPeersReady();
 			setTimeout(function () {
 				expect(peers.discover.calledOnce).to.be.ok;
-				peers.discover.restore();
 				done();
 			}, 100);
 		});

